docs(backups): encryption algorithms (#8266)
vatesfr/xen-orchestra: docs/docs/backups.md

![](./assets/log-runId.png)

## Backup Encryption

Xen Orchestra ensures robust data security for backups stored remotely, by leveraging advanced encryption algorithms. Here's a closer look at how encryption works and the technology behind it:

### Configuring encryption

Encryption is opt-in and requires configuring an encryption key on the remote.

1. Go to the Settings → Remote menu.

> 2. Go to the section called **New file system remote**, or edit an existing remote.

Reviewer: encryption can't be changed if a remote contains any backup (either enabling/disabling of changing the key )

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/4d784b0af9d1d060604fd4cde3ed8f00286342ee

3. In the subsection called **Encrypt all new data sent to this remote** you will find a text area. Enter your encryption key there.
4. Click the **Save configuration** Button to finish the encryption setup.

> 

Reviewer: encryption is only compatible with block based remote. (Full backup don't care, but incremental backup will fail if not )

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/4d784b0af9d1d060604fd4cde3ed8f00286342ee

### Current Encryption Algorithm: `AES-256-GCM`

> Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn't an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform. 

Reviewer: ```suggestion
Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn't an issue when working with incremental backups, as the data is split into smaller blocks, making it fully compatible with any remote ( S3 compatible or file based). 
Full backups creates one file per backup with all the data , that can go over 64GB, even when using xcp-ng zstd encryption
```

> In addition, this algorithm is fully compliant with [ANSSI guidelines (in French)](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).

Reviewer: this algorithm is authenticated, that means an additional metadata is put at the end of any saved file. This is used on restore to ensure that the data restored are exactly the data that were encrypted, thus detecting bit rot or a malicious attacker without the key. Note this is not a recoverable error, the file will be unusable
https://en.wikipedia.org/wiki/Authenticated_encryption

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/4d784b0af9d1d060604fd4cde3ed8f00286342ee

Author: ~~TODO : it's the same for the other algorithm~~

### Upcoming Change: `ChaCha20-Poly1305`

To improve flexibility and performance, Xen Orchestra will transition to the [`ChaCha20-Poly1305`](https://en.wikipedia.org/wiki/ChaCha20-Poly1305) encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.

> Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.

Reviewer: as aes-256-gcm , chacha20-poly1305 is authenticated

![](./assets/log-runId.png)

## Backup Encryption

> Xen Orchestra ensures robust data security for backups stored in Amazon S3 by leveraging advanced encryption algorithms. Here’s a closer look at how encryption works and the technology behind it:

Reviewer 2: Encryption is compatible will all remotes, not only Amazon S3.

Encryption is opt-in and requires configuring an encryption key on the remote (page _Settings/Remotes_), I think we should show that.

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/6587b3a0f9e384f5078433d77ac0d8e00724fbd4

### Current Encryption Algorithm: `AES-256-GCM`

> Currently, backups use the `AES-256-GCM` encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn’t an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform.

Reviewer 2: ```suggestion
Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn’t an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform.
```

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/259124a697cbda4162b4ad9daf33216e46186a35

> In addition, this algorithm is fully compliant with [ANSSI guidelines](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).

Reviewer 2: ```suggestion
In addition, this algorithm is fully compliant with [ANSSI guidelines 🇫🇷](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).
```

Author: Fixed by https://github.com/vatesfr/xen-orchestra/commit/e4be7238a4ab0da40c52f8db74497f96788a9535.

Replaced the flag with a 'French' label. Flags aren't good practice, as per https://www.flagsarenotlanguages.com/

### Upcoming Change: `ChaCha20-Poly1305`

> To improve flexibility and performance, Xen Orchestra will transition to the `ChaCha20-Poly1305` encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.

Reviewer 2: ```suggestion
To improve flexibility and performance, Xen Orchestra will transition to the [`ChaCha20-Poly1305`](https://en.wikipedia.org/wiki/ChaCha20-Poly1305) encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.
```

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/259124a697cbda4162b4ad9daf33216e46186a35

> Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.

Reviewer 2: ```suggestion
Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.

This algorithm is also fully compliant with [ANSSI guidelines 🇫🇷](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).
```

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/e4be7238a4ab0da40c52f8db74497f96788a9535.

Replaced the flag with a 'French' label. Flags aren't good practice, as per https://www.flagsarenotlanguages.com

### How It Works

#### AES-256-GCM

`AES-256-GCM` is a symmetric-key cryptographic block cipher, which is a method for encrypting data that ensures both confidentiality and integrity.

This method uses a 256-bit secret key and a unique initialization vector (IV) to encrypt data with the AES algorithm, transforming it into unreadable blocks.
The "GCM" part (Galois/Counter Mode) adds a verification tag to detect any tampering with the encrypted data. To decrypt, the same key and IV are required, and the tag ensures the data hasn’t been altered.

It’s a fast and secure system, commonly used in applications like secure web browsing and messaging.

#### ChaCha20-Poly1305

`ChaCha20-Poly1305` is a stream cypher, which is a method for encrypting messages.

Stream cyphers generate a random-looking sequence of bits, called a keystream, that matches the length of the message. This keystream is created using a secret key and an additional starting value called an initialization vector (IV).

To encrypt the message, the cipher combines each bit of the message with the corresponding bit of the keystream using a simple XOR operation (a basic "bit-flipping" process).

> The result is the encrypted message, or ciphertext. Modern stream ciphers are designed to ensure the keystream appears completely random, even if someone tries to manipulate the initialization vector.

Reviewer 2: While it is quite interesting, I don't think it makes sense to put in our documentation. :thinking:

Author: Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/afcdbd80c20608e4585b4fdf9df041c2c268423b

### Authentication

The encryption algorithms are authenticated, meaning additional metadata is appended to the end of each saved file. During restoration, this metadata ensures that the restored data matches the original encrypted data, allowing the system to detect issues like bit rot or tampering by an attacker without the encryption key. However, it's important to note that this is not a recoverable error—if the verification fails, the file will be unusable.

### Configuring encryption

Encryption is opt-in and requires configuring an encryption key on the remote.

:::warning
- Encryption is only compatible with block-based remotes.
- Encryption cannot be changed (such as enabling, disabling or changing the encryption key) if a remote contains any backup.
:::

1. Go to the Settings → Remote menu.
2. Go to the section called **New file system remote**, or edit an existing remote.
3. In the subsection called **Encrypt all new data sent to this remote** you will find a text area. Enter your encryption key there.
4. Click the **Save configuration** Button to finish the encryption setup.

> ### Current Encryption Algorithm: AES-256-GCM

Reviewer 2: - formatting
- change current/future to previous/current because the change will come rather soon, so that we don't have to update the doc in a week
- only keep the algo name in the title to make anchor links stable (i.e. not change in the future)
- explicit the date

```suggestion
### `AES-256-GCM`

> This algorithm was the default before February 2025 and has now been replaced by [`ChaCha20-Poly1305`](#chacha20-poly1305).
```

#### What is AES-256-GCM?

Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn't an issue when working with incremental backups, as the data is split into smaller blocks, making it fully compatible with any remote (S3-compatible or file-based). 

> Full backups create one file per backup with all the data, that can go over 64 GB, even when using XCP-ng zstd encryption.
> 
> 

Reviewer 2: ```suggestion

Full backups create one file per backup with all the data, that can go over 64 GB, even when using XCP-ng zstd encryption.

```

#### Compliance

The `AES-256-GCM` algorithm is fully compliant with [ANSSI guidelines (in French)](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).

> ### Upcoming Change: `ChaCha20-Poly1305`

Reviewer 2: Make anchor link more stable.

```suggestion
### `ChaCha20-Poly1305`
```

To improve flexibility and performance, Xen Orchestra will transition to the [`ChaCha20-Poly1305`](https://en.wikipedia.org/wiki/ChaCha20-Poly1305) encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.

> Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.

Reviewer 2: As it's now the new default, I believe this section should be above the legacy one.
